_posts/2017-09-22-redis-2-to-redis-3.md

```diff
@@ -0,0 +1,48 @@
+---
+layout:     post
+title:      Our journey from Redis 2 to Redis 3 while not taking the site down.
+date:       2017-09-22
+summary:    How we managed to upgrade from Redis 2 to Redis 3 without causing a site outage.
+category:   Devops
+tags:       redis
+author:     craig_stewart
+---
+We use Redis within Sky Betting & Gaming as a shared in memory cache for things like identity tokens that need to be known across api servers, or web servers. Within the Core Tribe this is used to help deal with the huge number of logins we have to handle day to day and particularly at busy times when we could have more than 20,000 people logging in within a single minute. This works well in so far as the data is readily available to a large number of servers (in the case of SSO tokens 70 Apache HTTPD servers). We've recently embarked upon a process of upgrading our redis servers, and this upgrade is leading us to use the native clustering features available from redis 3.2. This blog post hopes to explain why we're using clustering, what problems we have encountered along the way, and what our solutions have been.
+
+### In the beginning (or at least before the upgrade)
+
+Our legacy caches consisted of a pair of redis servers for each cache that we had, with keepalived running to ensure that there was always a master node, listening on a floating IP address. These failover pairs required considerable effort to manage when things went wrong, and the failure modes were sometimes quite interesting. On occasion the slave node, that would only allow reads of the data it held, and not writes, would end up with the floating IP address, which was relatively easy to diagnose, but broke whichever application was trying to use that cache at the time in painful ways.
+
+### The new application
+
+So whilst in this situation we needed to build a new application, one that used a shared in memory cache, but that we didn't want to be at the mercy of a dodgy failover process for that cache. So our requirements were a shared in memory cache, with no single point of failure, that could cope with multiple different failure modes using as little human intervention as possible, and also recover after those events cleanly, also with little human intervention, an additional ask was to improve the security of the cache to reduce the scope for data exfiltration (more on that later). At the time Redis Sentinel was looking promising, and there were a number of applications floating about to allow proxying of redis connections such as twemproxy. This lead to planning for a setup that would have many moving parts, it should have worked, with minimal human interaction, but it was complex and needed a large number of servers and services running and communicating with each other.
+
+![Redis Sentinel and TwemProxy](/images/Redis-Sentinel-and-TwemProxy.svg)
```

Comment:
This isn't scaling down well on my narrow screen.  I'm running at 1366 X 768 and some of the text is cut off.

Reply:
The graphic? It appears to scale reasonably on my machine using chrome, what browser are you using?

Reply:
Ah, it appears to be a Firefox issue.  It's scaling nicely in Chrome but in Firefox 55 things are getting cut off.
![screen shot 2017-09-25 at 10 44 24](https://user-images.githubusercontent.com/84906/30802484-98e00cd4-a1de-11e7-947f-c31a9f3f88d3.png)

Reply:
Looks like this can be solved by changing font size in the svg to 0.6em from 12px, it doesn't appear to change the render in chrome, but fixes it in firefox (on linux at least)